Add GetHWND() to wxWindowGTK when on Windows

ryancog opened the pull request:
As the title says, when using wxGTK on Windows there's not a way (that I can tell) to get an HWND from a wxWindow. This uses a couple of gdk and gtk helpers to get the HWND from the underlying GtkWidget.

ryancog committed:
Add GetHWND() to wxWindowGTK when on Windows

ryancog commented:
To give some context as to why I want this (although I'm sure there's plenty of other uses), in order to set the title bar to be dark when `GTK_CSD=0` (And I always turn CSD off... it's awful under windows), the window's HWND is required:

```c++
#	ifdef __WINDOWS__
    DWORD useDarkMode{Misc::getTheme() == Misc::Theme::DARK_MODE};
    DwmSetWindowAttribute(
            GetHWND(),
            DWMWINDOWATTRIBUTE::DWMWA_USE_IMMERSIVE_DARK_MODE,
            &useDarkMode,
            sizeof(DWORD)
            );
#	endif
```

vadz commented on src/gtk/window.cpp:
```diff
@@ -4240,6 +4243,14 @@ bool wxWindowGTK::GTKShowFromOnIdle()
 }
 
+#ifdef __WINDOWS__
+WXHWND wxWindowGTK::GetHWND() const {
```
Very minor, but wx style is to put braces on their own lines, i.e.
```suggestion
WXHWND wxWindowGTK::GetHWND() const
{
```

ryancog replied:
Whoops, my bad... I didn't realize the difference in styling it seems.

vadz commented on include/wx/gtk/window.h:
```diff
@@ -139,6 +139,9 @@ class WXDLLIMPEXP_CORE wxWindowGTK : public wxWindowBase
 
     virtual WXWidget GetHandle() const override { return m_widget; }
+#ifdef __WINDOWS__
+    WXHWND GetHWND() const;
```
I'm not sure if it's a good or bad idea to use the same name as for the existing function in wxMSW. On one hand, this would allow referring to it without checking for `__WXGTK__` but OTOH do we really plan to have much code using either wxMSW or wxGTK and having to manipulate the underlying `HWND` directly?

Perhaps we should call this `GTKGetWin32Handle()` instead? Granted, this is much longer but this shouldn't be used very often, hopefully.

ryancog replied:
> but OTOH do we really plan to have much code using either wxMSW or wxGTK and having to manipulate the underlying `HWND` directly?

I can't imagine so.

> Perhaps we should call this `GTKGetWin32Handle()` instead? Granted, this is much longer but this shouldn't be used very often, hopefully.

Making it a different, distinguishable name makes sense to me.

Mostly out of curiousity, what's the logic stylistically behind prefixing it with `GTK`?
Since there's other wxWindowGTK functions that aren't...

vadz commented on src/gtk/window.cpp:
```diff
@@ -4240,6 +4243,14 @@ bool wxWindowGTK::GTKShowFromOnIdle()
 
+#ifdef __WINDOWS__
+WXHWND wxWindowGTK::GetHWND() const {
+    // There isn't an underlying HWND if the widget hasn't been realized yet.
+    gtk_widget_realize(m_widget);
```
This is more serious, IMO it's really unexpected for an accessor like this to have significant side effects.

Would it be a problem to just call this function only if the window had been already realized?

ryancog replied:
Would need to add an extra check on the first gtk function, since it'll return null if not (and passing that to gdk will cause *bad things*), but that's easy enough.

However, that'd require a way (for my use case at least) to be able to realize the GTK Window through wxWidgets. I'll see if there's a function for that already. If not, I'd want to add one.

ryancog replied:
I don't see a method to "force" realize a gtk window.

Are you aware if there's something like this that exists and/or what sort of side effects would need to be considered by creating another function (GTKRealizeWidget() for example) to do this?

vadz reviewed (commented):
Thanks, I don't think there is a problem with adding such a function but I'm not sure about the name and it should also be documented in `interface/wx/window.h` if we mean it to be part of the public API now.

ryancog committed:
Fix wxWindowGTK::GetHWND() styling

Referencing https://github.com/wxWidgets/wxWidgets/pull/24675#discussion_r1666109964

Co-authored-by: VZ <vz-github@zeitlins.org>

ryancog commented:
> it should also be documented in interface/wx/window.h if we mean it to be part of the public API now.

Should it go in `interface/wx/window.h` or into a gtk specific one? That file currently seems to be entirely platform-agnostic functions, so I'm not sure how to go about adding that documentation.

ryancog committed:
Rename wxWindowGTK::GetHWND() to GTKGetWin32Handle(), add GTKRealizeWidget()

https://github.com/wxWidgets/wxWidgets/pull/24675

vadz commented:
Thanks for the changes!

@paulcor Do you mind adding `GTKRealizeWidget()`? I'm not totally sure if it's a good idea, but OTOH I don't have any better ones.

paulcor commented:
A shown window is realized, it might be sufficient just to require that rather than adding `GTKRealizeWidget()`.

ryancog commented:
That feels a bit hacky... especially if I don't want to show it yet. Would `Show(); Hide();` work? Even then it isn't the most obvious...

paulcor commented:
`wxWindowCreateEvent` is generated when a window is realized, things requiring the `GdkWindow` can probably be done from a handler for that event.

ryancog commented:
Ok, that seems reasonable then. Something I think would be reasonable to note writing documentation for `GTKGetWin32Handle()`. 

Speaking of which, I’m still not entirely sure where to put such documentation.

vadz commented:
> Speaking of which, I’m still not entirely sure where to put such documentation.

In `interface/wx/window.h`, please. TIA!

ryancog committed:
Remove GTKRealizeWidget(), add documentation for GTKGetWin32Handle()

See https://github.com/wxWidgets/wxWidgets/pull/24675

ryancog commented:
Ok, I've removed `GTKRealizeWidget()`, and added documentation for `GTKGetWin32Handle()`.

Let me know if there's anything wrong with that documentation... Not sure if it's overly verbose or anything along those lines, and I'm unsure if there's extra formatting I should do. (Closest thing to doxygen I've used is JavaDocs... and even then it was only lightly)

ryancog committed:
Remove GTKRealizeWidget() from include/wx/gtk/window.h

ryancog commented:
Bumping this.

If there's no issues with the documentation or anything else, I don't have anything more to add.

vadz commented:
Sorry for the delay, will merge this soon.
